### ISSUE TITLE
Links with non-HTTP URL schemes (e.g. mailto:) cause crawl to crash

### ISSUE THREAD

**Issue by alexwlchan** [inglesp/http-crawler#6]
Steps to repro:
1. Create a file called `index.html` with the following contents:
   
   ``` html
   <html>
       <body>
           <a href="mailto:example@example.org">Email me!</a>
       </body>
   </html>
   ```
2. Within the same directory as that file, start a web server for http-crawler to crawl:
   
   ```
   python -m SimpleHTTPServer
   ```
   
   If you open http://localhost:8000 in a web browser, you should see "Email me!".
3. Use the example from the README to scrape this mini-site:
   
   ``` pycon
   >>> for rsp in crawl('http://localhost:8000'):
   ...     if rsp.status_code != 200:
   ...         print('Got {} at {}'.format(rsp.status_code, rsp.url))
   ...
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
     File "/Users/alexwlchan/.virtualenvs/tempenv-31a63173800c1/lib/python3.5/site-packages/http_crawler/__init__.py", line 22, in crawl
       rsp = session.get(url)
     File "/Users/alexwlchan/.virtualenvs/tempenv-31a63173800c1/lib/python3.5/site-packages/requests/sessions.py", line 487, in get
       return self.request('GET', url, **kwargs)
     File "/Users/alexwlchan/.virtualenvs/tempenv-31a63173800c1/lib/python3.5/site-packages/requests/sessions.py", line 475, in request
       resp = self.send(prep, **send_kwargs)
     File "/Users/alexwlchan/.virtualenvs/tempenv-31a63173800c1/lib/python3.5/site-packages/requests/sessions.py", line 579, in send
       adapter = self.get_adapter(url=request.url)
     File "/Users/alexwlchan/.virtualenvs/tempenv-31a63173800c1/lib/python3.5/site-packages/requests/sessions.py", line 653, in get_adapter
       raise InvalidSchema("No connection adapters were found for '%s'" % url)
   requests.exceptions.InvalidSchema: No connection adapters were found for 'mailto:example@example.org'
   ```

This will probably affect other URL schemes, e.g. `ftp://` and `file://`, but I haven’t tried those.

I don’t think it’s completely obvious what the correct behaviour should be. Crashing like this is sub-optimal, but silently ignoring the URL might not be a good thing either – what if I have a mistyped URL like `htp://example.net`? Might that be useful to find?

If somebody’s who’s making their first contribution to Open Source would like to try fixing this, I’d be happy to offer help and/or advice.

I think a good start would be to suggest what http-crawler should do when it encounters one of these URLs, and then we can discuss how you can go about fixing it.

**Comment by inglesp** [comment]
Hi @alexwlchan, thanks for this!

There are three kinds of scheme we're interested in:
- `http` and `https`, which can be handled by requests.  There's nothing to do here.
- Valid schemes that can't be handled by requests.  My gut feeling is that these should simply be ignored for now.
- Invalid schemes.  These should be communicated to the caller somehow, as they indicate a possible error, but it's not obvious how.  As things stand, `http_crawler.crawl` is expected to yield `Response` objects, so yielding anything else would be surprising.  Perhaps instead, `http_crawler.crawl` should yield some kind of object with attributes like `.url`, `.response`, and for this case, `.error`.  What do you think?

As for valid and invalid schemes, I've found [this list](http://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml) which could be used.

**Comment by alexwlchan** [comment]
Ignoring valid schemes that requests can’t handle seems sensible.

For the invalid schemes, returning something that looks like a `Response` seems possible. Perhaps subclass the `Response` class from requests? That way you get an API that’s compatible with whatever version of requests you’re using, and then you add a `.error` attribute to provide http-crawler specific information.

**Comment by alexwlchan** [comment]
Here’s another URL that causes it to crash:

```
requests.exceptions.InvalidSchema: No connection adapters were found for 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIIAAAAoCAMAAAAFWtJHAAAAwFBMVEUAAAAAAAAAAABtbW0AAAAAAAAAAAAAAAAAAAAAAAClpaUAAADp6ekAAAD5+fna2toAAAAMDAzv7+/Nzc0AAAA2Njb8/Pz9/f3T09MAAAAAAAD7+/sAAAArKyuxsbH39/fs7OwbGxuIiIjz8/N8fHyenp7u7u74+PgAAAC8vLxWVlbx8fF1dXXl5eVcXFyUlJTQ0NDFxcVCQkLAwMC4uLj19fXo6OjW1tarq6ve3t77+/vi4uL6+vrKysrNzc3///8w7gSSAAAAP3RSTlMAOQNdPSYBPywKexLLGPCxNEHXnzFL+v2nGwf1IEiE6dBFad9jd9PuLo1V2mDDV3Cjl06SiuXIq4C3973ym6BQMVUPAAAEXElEQVR4Ae2WCVP6OBiH05L0l1IqrVbkKHJ54I0oHn+PfP9vtUle0z/YdhbH2XVnd58ZnRJIeHiPJOx//mH4vQSAN+8FjAhFxgHIaPvJeZ99hxwEElon5iAQbj85Y98g8ODwjEOMAvGFyeE3FEKgodTBqj0BJGN9DhyNd5Ta3ean9QEopfaA+LsKhnEKRExqg4FSP6Og7oEkAjBWnxSCgBX4xF+kcLoPcOBQrSv0e5kH7s1j37jECQieCTPiFGxL5VHw2zQWCeeJiPt6kjRQw0XSkIdVChf67xGa4alSnZlT6HEQ8CK9ANbhvXUF9xlDkBfTuHDWScgC9+z5FQpPI12TlwC6+sV7ixR8CUMKiwjm2GQeOQWHMGuHGdbnObJAwCEqFJpNU5H6uaPUaEIKiQfg+PHk1+u4OwW9PlWW2ctbA4BHCtp+cNK+H8Jos4gDmC5ar4Nx9waaG/2B13NgDqS7+vm2RgEtEws82P+kwIHhs/pgkQKcFIhfd7CogtGNjYMHTLpurD0ERbYFw4JaD3GlQuNAL/JEsSAF4HqlCnaHACk4WhOn4OgCkMD5hSpYNYDJTD8Y46n+jsE1kPhVCuR6QBXhFK7MUOu9O6b1SWF3b+/9ZVWMGOlu93E8UDaAhgc7bfH+0DHqKXCkHzoNDFfU+zxiVQrUC9QXTuHYtKpN59OA3IxCG4b7jh6ZFuVockaNTW09mkJzOaPU49a6mE9cAchZpQJNpUWcwgV9r6FJswsFKrITp2B5pMBMdnS0z2HZNy2+BNKxSZxZfglkrFYBJxQnpzA5sN/HheR2aFQoZBLAi149dQoyAYYjW0hHlHguBAdMcR0DuDZ5omevX6+AI8qcU7ikKT3GBHCnXwydgmCC0tRwCnGQ2Wp6Be71yNIWfQSkOl9vAI1SBCNWrwC01RROgX7BuT2HI4r7tFAw086p/NwZEdOEa7R1uAFuNmQPuKAEAjYNQ0CyeoUEWHYBnpQVQgpvc0Ph+gsKlAnKg1+vEHsw5LKciLKCAJobiWBzYFGbCKpHqkZZrxBFHEASyFI59vJPCskcwNVGOWZAOqsrR+pKbaNeAMT1CixMEtlnsqopNxUMzVJT3tY35aXZm6a6Y9QhwMN6BUJWbE1lhbMO1WehkO7poO0sK7em9MJGxp1XSbC1gtugzzSLQmGsX7VntJGSwsPZ2d2z3bIPKzdoOp3Wzqt8G4XyMVUoFIxLx1S7+piaHtCvR3FeRVsq0GFdp9C5TbGpcNqsPqyHKxcfd14h21KhuLKUFU4f3osrC7F6uV3WXFnadL7wyAPeKDXw2RoJCO5GY4DouYvb/gepVXheLoewzPseQG9N/vzilrMIjoStE3++zvle4eSurw7XEe76ynI4aq+v7lEyt1x5awiFlFLQbHKIpabnM3eJLym4Szzzc/du7SU+zOXv9UNpECH7IoH/gecURPlN9vdQpeD47yhIFNX0U0QgvID9nENm+yxk/xb+AGAjNfRZuk9qAAAAAElFTkSuQmCC’
```

**Comment by mei-li** [comment]
Hi @alexwlchan and @inglesp I can look into the issue. I am not sure whether I would like that it crashes the crawling when a non valid scheme is encountered either, maybe a warning message is enough. Anyways @inglesp do you expect that the valid scheme list is copied in the code, or dynamically looked up?

**Comment by inglesp** [comment]
Hi @mei-li 

Thanks for taking a look!

1) Emitting a warning is preferable to crashing, but it'd be nice if this was configurable.  What about a `crash_on_invalid_url_scheme` argument to `crawl`?
2) I think it'd be better to have a valid scheme list that was hardcoded into the library, rather than looking the list up on every run.
2a) It would be great to have a script that would update the hardcoded list by scraping http://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml.

What do you think?

**Comment by alexwlchan** [comment]
> It would be great to have a script that would update the hardcoded list by scraping http://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml

I am definitely 👍 on having a script to update it, although note that the same page provides [a CSV file](https://www.iana.org/assignments/uri-schemes/uri-schemes-1.csv) which is probably much easier to parse.

**Comment by inglesp** [comment]
Basically fixed by #12.
